pkg/framework/coremodel/registry/provide.go

```diff
@@ -1,40 +1,59 @@
 package registry
 
 import (
+	"sync"
+
+	"github.com/google/wire"
+	"github.com/grafana/grafana/pkg/cuectx"
+	"github.com/grafana/grafana/pkg/framework/coremodel"
 	"github.com/grafana/thema"
 )
 
-// ProvideStatic provides access to individual coremodels via explicit method calls.
-//
-// Prefer this to the ProvideGeneric type when your code works with known,
-// specific coremodels(s), rather than generically across all of them. This
-// allows standard Go static analysis tools to determine which code is depending
-// on particular coremodels.
+// CoremodelSet contains all of the wire-style providers related to coremodels.
+var CoremodelSet = wire.NewSet(
+	NewBase,
+)
+
+// NewBase provides a registry of all coremodels, without any composition of
+// plugin-defined schemas.
 //
-// This will use the default Grafana thema.Library, defined in pkg/cuectx, which
-// will avoid duplicate parsing of Thema CUE schemas. If you need control over the
-// thema.Library in use, use ProvideStaticWithLib instead.
-func ProvideStatic() (*Static, error) {
-	return provideStatic(nil)
+// The returned registry will use the default Grafana thema.Library, defined in
+// pkg/cuectx. If you need control over the thema.Library used by the coremodel
+// lineages, use NewBaseWithLib instead.
+func NewBase() *Base {
+	return provideBase(nil)
 }
 
-// ProvideStaticWithLib is the same as ProvideStatic, but
-// allows control over the thema.Library used to initialize the underlying
-// coremodels.
+// NewBaseWithLib is the same as NewBase, but allows control over the
+// thema.Library used to initialize the underlying coremodels.
 //
-// Prefer ProvideStatic unless you absolutely need this control.
-func ProvideStaticWithLib(lib thema.Library) (*Static, error) {
-	return provideStatic(&lib)
+// Prefer NewBase unless you absolutely need this control.
+func NewBaseWithLib(lib thema.Library) *Base {
+	return provideBase(&lib)
 }
 
-// ProvideGeneric provides a simple Generic registry of all coremodels.
-//
-// Prefer this to the static ProvideStatic when your code needs to
-// work with all coremodels generically, rather than specific coremodels.
-func ProvideGeneric() (*Generic, error) {
-	return provideGeneric()
+var (
+	baseOnce    sync.Once
+	defaultBase *Base
+)
+
+func provideBase(lib *thema.Library) *Base {
+	if lib == nil {
+		baseOnce.Do(func() {
+			defaultBase = doProvideBase(cuectx.ProvideThemaLibrary())
+		})
+		return defaultBase
+	}
+
+	return doProvideBase(*lib)
 }
 
-// NOTE - no ProvideRegistryWithLib is defined because there are no anticipated
-// cases where a caller would need to operate generically across all coremodels,
-// and control the library they're initialized with. If that changes, add one.
+// All returns a slice of all registered coremodels.
+//
+// Prefer this method when operating generically across all coremodels.
+//
+// The returned slice is sorted lexicographically by coremodel name. It should
+// not be modified.
```

Comment:
I've often thought similar things. Sad truth is that, afaict, this is often considered enough.

I usually get pretty particular about these things. the only reason I didn't here is because I know how I actually want to do it in this case, but decided to defer until later: return a codegen'd  array with each field of the registry struct literally mapped into place lexicographically.

Should then be just one alloc per call, no need for sorting at runtime, and no worries if a caller futzes with order of their returned slice.